### ISSUE TITLE
Is this really supposed to always be enabled?

### ISSUE THREAD

**Issue by aaronmondal** [AdaptiveCpp/AdaptiveCpp#1046]
The `+sm_60` seems a bit odd to me. Is this just an artifact from development or is this enabling some specific feature that's always required? If it is intended, a comment above the line might be helpful.

https://github.com/OpenSYCL/OpenSYCL/blob/12fdcaedfa990ab58ddf8bce304fa8cf917e6182/src/libkernel/sscp/ptx/CMakeLists.txt#L7

**Comment by illuhad** [comment]
AFAIR That was needed to make clang expose some nvvm atomic builtins that we need in our bitcode libraries, which are only available on sm_60 and newer (see what happens when you remove it :-) ). On older hardware, as long as the respective atomic functions are not used, it will still work thanks to our SSCP late kernel outlining magic. But clang is not really designed for what we do, where there's potentially different hardware generations targeted in a single LLVM IR module and target specialization is done as part of control flow and handled using dead code elimination passes at JIT time. So we have to resort to solutions like those, which admittedly don't look nice.

Yeah, maybe a comment is a good idea.

**Comment by aaronmondal** [comment]
Ok that makes sense. I don't have access to an NVPTX device ATM, so I'll just take your word for it and copy it into our build files :laughing:

Out of curiosity: Variations of LLVM IR always make me think of potential use cases for MLIR. During the design of SSCP, did this come up in discussions, or is the SSCP logic "too custom" to fit into something like that?

How deeply are the llvm-to backends tied to SYCL, or rather, would it make sense to try to adapt LLVM in a way that makes such a compilation flow useable for other heterogeneous compilation flows? Considering how much compile time this saves there might be a broader interest in the LLVM community to have such logic in upstream LLVM.

**Comment by illuhad** [comment]
> Ok that makes sense. I don't have access to an NVPTX device ATM, so I'll just take your word for it and copy it into our build files laughing

If you remove it, you would already notice the problem when building Open SYCL as long as the CUDA backend is enabled - regardless of whether you have an NVIDIA device or not ;)

> Out of curiosity: Variations of LLVM IR always make me think of potential use cases for MLIR. During the design of SSCP, did this come up in discussions, or is the SSCP logic "too custom" to fit into something like that?

It's unclear to what extent we could benefit from MLIR. It's not really that the LLVM IR that we use has different IR variants, I would rather describe the issue as the clang frontend being too restrictive w.r.t which builtins it exposes. Ultimately it's just a function call, and the question is just whether clang declares that function.
The big issue with MLIR that I'm hearing from the folks at the SYCL-MLIR project is the C++ compiler frontend, which contains multiple restrictions, and they have to perform some heroic hacks to get the host code to work (those restrictions are generally not an issue for device code, where some C++ functionality like exceptions is not allowed anyway). I would expect that for a single-pass compiler like ours where host and device code are even more tightly coupled, these issues would be exacerbated.

> How deeply are the llvm-to backends tied to SYCL, or rather, would it make sense to try to adapt LLVM in a way that makes such a compilation flow useable for other heterogeneous compilation flows? Considering how much compile time this saves there might be a broader interest in the LLVM community to have such logic in upstream LLVM.

The coupling to SYCL itself is loose, as the majority of our SSCP compiler stuff lives at the LLVM IR level, not the clang level.
I don't know though if upstream would be willing to accept our code, since LLVM IR was not really made to be jitted to multiple different backends.
This requires changing the LLVM target triplet and datalayout in the LLVM module to conform to the expectations of various LLVM GPU backends. And this in turn can break the IR at some places, and requires fixing some things in the IR, such as changed default alloca address spaces and similar. All of that could be perceived as hacks ;)
The reason that changing datalayout is in principle safe in our case and not as bad as it sounds is that, since SYCL is a single-source model, we already *know* that at least the basics such as layout of data structures and size of data types *must* be compatible between host and device code (otherwise everything will break down) and so we know in advance that datalayout changes from host code to GPU won't break IR at a fundamental level. For separate source models like OpenCL, things might be different.
So in summary, SYCL coupling is loose, but there are some other assumptions.
